File: jsk_2015_06_hrp_drc/drc_task_common/src/drc_task_common/drc_teleop_interface.cpp
#include "rviz/config.h"
#include "drc_teleop_interface.h"
#include "ros/time.h"
#include <ros/package.h>
#include "ui_drc_teleop_interface.h"
#include "dynamic_reconfigure/Reconfigure.h"

using namespace rviz;
namespace drc_task_common
{
  DRCTeleopInterfaceAction::DRCTeleopInterfaceAction( QWidget* parent )
    : rviz::Panel( parent )
  {
    ui_ = new Ui::DRCTeleopInterface();
    ui_->setupUi(this);
    ui_->verticalLayout->setAlignment(Qt::AlignLeft);

    ros::NodeHandle nh("~");
    
    ui_->reset_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->reset_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/reset-pose.jpg")).c_str()))));
    ui_->reset_manip_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->reset_manip_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/reset-manip-pose.jpg")).c_str()))));
    ui_->drill_reset_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->drill_reset_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/drill-reset-pose.jpg")).c_str()))));
    ui_->watch_drill_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->watch_drill_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/watch-drill-pose.jpg")).c_str()))));
    ui_->door_through_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->door_through_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/door-through-pose.png")).c_str()))));
    ui_->hand_reset_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hand_reset_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hand-reset-pose.jpg")).c_str()))));
    ui_->hand_hook_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hand_hook_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hand-hook-pose.jpg")).c_str()))));
    ui_->hand_hook_pose_after_5sec_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hand_hook_pose_after_5sec_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hand-hook-pose.jpg")).c_str()))));
    ui_->hand_grasp_pose_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hand_grasp_pose_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hand-grasp-pose.jpg")).c_str()))));
    ui_->hrpsys_hand_calib_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_hand_calib_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hand-calib.png")).c_str()))));
    ui_->hrpsys_start_abc_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_start_abc_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/start-abc.png")).c_str()))));
    ui_->hrpsys_start_st_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_start_st_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/start-st.png")).c_str()))));
    ui_->hrpsys_start_imp_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_start_imp_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/start-imp.png")).c_str()))));
    ui_->hrpsys_start_imp_soft_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_start_imp_soft_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/start-imp-soft.png")).c_str()))));
    ui_->hrpsys_start_imp_hard_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_start_imp_hard_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/start-imp-hard.png")).c_str()))));
    // ui_->hrpsys_stop_abc_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    // ui_->hrpsys_stop_abc_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/stop-abc.png")).c_str()))));
    // ui_->hrpsys_stop_st_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    // ui_->hrpsys_stop_st_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/stop-st.png")).c_str()))));
    ui_->hrpsys_stop_imp_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hrpsys_stop_imp_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/stop-imp.png")).c_str()))));
    ui_->display_manip_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->display_manip_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/display_6dof.png")).c_str()))));
    ui_->hide_manip_button->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    ui_->hide_manip_button->setIcon(QIcon(QPixmap(QString((ros::package::getPath("drc_task_common")+std::string("/icons/hide_6dof.png")).c_str()))));
    connect( ui_->reset_pose_button, SIGNAL( clicked() ), this, SLOT( callRequestResetPose()));
    connect( ui_->reset_manip_pose_button, SIGNAL( clicked() ), this, SLOT( callRequestResetManipPose()));
    connect( ui_->drill_reset_pose_button, SIGNAL( clicked() ), this, SLOT( callRequestDrillResetPose()));
    connect( ui_->watch_drill_pose_button, SIGNAL( clicked() ), this, SLOT( callRequestWatchDrillPose()));
    connect( ui_->door_through_pose_button, SIGNAL( clicked() ), this, SLOT( callRequestDoorThroughPose()));

    connect( ui_->hand_reset_pose_button, SIGNAL( clicked() ), this, SLOT(  callRequestResetGripperPose()));
    connect( ui_->hand_hook_pose_button, SIGNAL( clicked() ), this, SLOT(  callRequestHookGrippePose()));
    connect( ui_->hand_hook_pose_after_5sec_button, SIGNAL( clicked() ), this, SLOT(  callRequestHookGrippePoseAfter5sec()));
    connect( ui_->hand_grasp_pose_button, SIGNAL( clicked() ), this, SLOT(  callRequestGraspGrippePose()));

    connect( ui_->hrpsys_start_abc_button, SIGNAL( clicked() ), this, SLOT(  callRequestStartABC()));
    connect( ui_->hrpsys_start_st_button, SIGNAL( clicked() ), this, SLOT(  callRequestStartST()));
    connect( ui_->hrpsys_start_imp_button, SIGNAL( clicked() ), this, SLOT(  callRequestStartIMP()));
    connect( ui_->hrpsys_start_imp_soft_button, SIGNAL( clicked() ), this, SLOT(  callRequestStartIMPSoft()));
    connect( ui_->hrpsys_start_imp_hard_button, SIGNAL( clicked() ), this, SLOT(  callRequestStartIMPHard()));

    // connect( ui_->hrpsys_stop_abc_button, SIGNAL( clicked() ), this, SLOT(  callRequestStopABC ()));
    // connect( ui_->hrpsys_stop_st_button, SIGNAL( clicked() ), this, SLOT(  callRequestStopST()));
    connect( ui_->hrpsys_stop_imp_button, SIGNAL( clicked() ), this, SLOT(  callRequestStopIMP()));

    connect( ui_->hrpsys_hand_calib_button, SIGNAL( clicked() ), this, SLOT(  callRequestHandCalib()));

    connect( ui_->display_manip_button, SIGNAL( clicked() ), this, SLOT(  callRequestDisplayManip ()));
    connect( ui_->hide_manip_button, SIGNAL( clicked() ), this, SLOT(  callRequestHideManip ()));

  }

  void DRCTeleopInterfaceAction::callRequestResetPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::RESET_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestResetManipPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::RESET_MANIP_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestDrillResetPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::DRILL_RESET_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestWatchDrillPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::WATCH_DRILL_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestDoorThroughPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::DOOR_THROUGH_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestResetGripperPose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HAND_RESET_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestHookGrippePose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HAND_HOOK_POSE);
  };
  void DRCTeleopInterfaceAction::callRequestHookGrippePoseAfter5sec(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HAND_HOOK_AFTER_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestGraspGrippePose(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HAND_GRASP_POSE);
  };

  void DRCTeleopInterfaceAction::callRequestHandCalib(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::CALIB_HAND);
  };

  void DRCTeleopInterfaceAction::callRequestStartABC(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_START_ABC);
  };

  void DRCTeleopInterfaceAction::callRequestStartST(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_START_ST);
  };

  void DRCTeleopInterfaceAction::callRequestStartIMP(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_START_IMP);
  };

  void DRCTeleopInterfaceAction::callRequestStartIMPSoft(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_START_IMP_SOFT);
  };

  void DRCTeleopInterfaceAction::callRequestStartIMPHard(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_START_IMP_HARD);
  };

  // void DRCTeleopInterfaceAction::callRequestStopABC(){
  //   callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_STOP_ABC);
  // };

  // void DRCTeleopInterfaceAction::callRequestStopST(){
  //   callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_STOP_ST);
  // };

  void DRCTeleopInterfaceAction::callRequestStopIMP(){
    callRequestUint8Request(drc_com_common::OCS2FCSmall::HRPSYS_STOP_IMP);
  };

  void DRCTeleopInterfaceAction::callRequestDisplayManip(){
    ros::ServiceClient client = nh_.serviceClient<dynamic_reconfigure::Reconfigure>("/transformable_interactive_server/set_parameters", true);
    dynamic_reconfigure::Reconfigure srv;
    dynamic_reconfigure::BoolParameter bool_param;
    bool_param.name = std::string("display_interactive_manipulator");
    bool_param.value = true;
    srv.request.config.bools.push_back(bool_param);
    if(client.call(srv))
      {
        ROS_INFO("Call Success");
      }
    else{
      ROS_ERROR("Service call FAIL");
    };
  };

  void DRCTeleopInterfaceAction::callRequestHideManip(){
    ros::ServiceClient client = nh_.serviceClient<dynamic_reconfigure::Reconfigure>("/transformable_interactive_server/set_parameters", true);
    dynamic_reconfigure::Reconfigure srv;
    dynamic_reconfigure::BoolParameter bool_param;
    bool_param.name = std::string("display_interactive_manipulator");
    bool_param.value = false;
    srv.request.config.bools.push_back(bool_param);
    if(client.call(srv))
      {
        ROS_INFO("Call Success");
      }
    else{
      ROS_ERROR("Service call FAIL");
    };
  };

  void DRCTeleopInterfaceAction::callRequestUint8Request(uint type){
    ros::ServiceClient client = nh_.serviceClient<drc_task_common::Uint8Request>("/uint8_command", true);
    drc_task_common::Uint8Request srv;
    srv.request.type = type;
    if(client.call(srv))
      {
        ROS_INFO("Call Success");
      }
    else{
      ROS_ERROR("Service call FAIL");
    };
  }

  void DRCTeleopInterfaceAction::save( rviz::Config config ) const
  {
    rviz::Panel::save( config );
  }

  void DRCTeleopInterfaceAction::load( const rviz::Config& config )
  {
    rviz::Panel::load( config );
  }
}

#include <pluginlib/class_list_macros.h>
PLUGINLIB_EXPORT_CLASS(drc_task_common::DRCTeleopInterfaceAction, rviz::Panel )

File: jsk_2015_06_hrp_drc/drc_task_common/src/drc_task_common/drc_teleop_interface.py
import os
import rospy
import rospkg

from qt_gui.plugin import Plugin
from python_qt_binding import loadUi
import python_qt_binding.QtCore as QtCore
from python_qt_binding.QtGui import *
from drc_task_common.srv import EusCommand

class DRCTeleopInterface(Plugin):

    def __init__(self, context):
        super(DRCTeleopInterface, self).__init__(context)
        # Give QObjects reasonable names
        self.setObjectName('drc_teleop_interface')

        # Process standalone plugin command-line arguments
        from argparse import ArgumentParser
        parser = ArgumentParser()
        # Add argument(s) to the parser.
        parser.add_argument("-q", "--quiet", action="store_true",
                      dest="quiet",
                      help="Put plugin in silent mode")
        args, unknowns = parser.parse_known_args(context.argv())
        if not args.quiet:
            print('arguments: ', args)
            print('unknowns: ', unknowns)

        # Create QWidget
        self._widget = QWidget()
        # Get path to UI file which should be in the "resource" folder of this package
        ui_file = os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'config', 'drc_teleop_interface.ui')
        # Extend the widget with all attributes and children from UI file
        loadUi(ui_file, self._widget)
        # Give QObjects reasonable names
        self._widget.setObjectName('drc_teleop_interfaceUi')
        # Show _widget.windowTitle on left-top of each plugin (when 
        # it's set in _widget). This is useful when you open multiple 
        # plugins at once. Also if you open multiple instances of your 
        # plugin at once, these lines add number to make it easy to 
        # tell from pane to pane.
        if context.serial_number() > 1:
            self._widget.setWindowTitle(self._widget.windowTitle() + (' (%d)' % context.serial_number()))
        # Add widget to the user interface
        context.add_widget(self._widget)

        # ROS service client
        global eus_commmand_srv
        eus_commmand_srv = rospy.ServiceProxy('/eus_command', EusCommand)

        # Robot Pose Button
        ## Reset Pose
        self._widget.reset_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'reset-pose.jpg'))))
        self._widget.reset_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.reset_pose_button.clicked[bool].connect(self.reset_pose_cb)
        ## Reset Manip Pose
        self._widget.reset_manip_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'reset-manip-pose.jpg'))))
        self._widget.reset_manip_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.reset_manip_pose_button.clicked[bool].connect(self.reset_manip_pose_cb)
        ## Init Pose
        self._widget.init_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'init-pose.jpg'))))
        self._widget.init_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.init_pose_button.clicked[bool].connect(self.init_pose_cb)

        # Hand Pose Button
        ## Reset Pose
        self._widget.hand_reset_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'hand-reset-pose.jpg'))))
        self._widget.hand_reset_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hand_reset_pose_button.clicked[bool].connect(self.hand_reset_pose_cb)
        ## Hook Pose
        self._widget.hand_hook_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'hand-hook-pose.jpg'))))
        self._widget.hand_hook_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hand_hook_pose_button.clicked[bool].connect(self.hand_hook_pose_cb)
        ## Grasp Pose
        self._widget.hand_grasp_pose_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'hand-grasp-pose.jpg'))))
        self._widget.hand_grasp_pose_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hand_grasp_pose_button.clicked[bool].connect(self.hand_grasp_pose_cb)

        # Hrpsys Button
        ## Start ABC
        self._widget.hrpsys_start_abc_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'start-abc.png'))))
        self._widget.hrpsys_start_abc_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_start_abc_button.clicked[bool].connect(self.hrpsys_start_abc_cb)
        ## Start ST
        self._widget.hrpsys_start_st_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'start-st.png'))))
        self._widget.hrpsys_start_st_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_start_st_button.clicked[bool].connect(self.hrpsys_start_st_cb)
        ## Start IMP
        self._widget.hrpsys_start_imp_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'start-imp.png'))))
        self._widget.hrpsys_start_imp_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_start_imp_button.clicked[bool].connect(self.hrpsys_start_imp_cb)
        ## Stop ABC
        self._widget.hrpsys_stop_abc_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'stop-abc.png'))))
        self._widget.hrpsys_stop_abc_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_stop_abc_button.clicked[bool].connect(self.hrpsys_stop_abc_cb)
        ## Stop ST
        self._widget.hrpsys_stop_st_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'stop-st.png'))))
        self._widget.hrpsys_stop_st_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_stop_st_button.clicked[bool].connect(self.hrpsys_stop_st_cb)
        ## Stop IMP
        self._widget.hrpsys_stop_imp_button.setIcon(QIcon(QPixmap(os.path.join(rospkg.RosPack().get_path('drc_task_common'), 'icons', 'stop-imp.png'))))
        self._widget.hrpsys_stop_imp_button.setToolButtonStyle(QtCore.Qt.ToolButtonTextUnderIcon)
        self._widget.hrpsys_stop_imp_button.clicked[bool].connect(self.hrpsys_stop_imp_cb)

    # Robot Pose Button Callback
    def reset_pose_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :angle-vector (send *robot* :reset-pose) 5000)')
        print('send reset_pose command')
    def reset_manip_pose_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :angle-vector (send *robot* :reset-manip-pose) 5000)')
        print('send reset_manip_pose command')
    def init_pose_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :angle-vector (send *robot* :init-pose) 5000)')
        print('send init_pose command')

    # Hand Pose Button Callback
    def hand_reset_pose_cb(self, checked):
        eus_commmand_srv(command='(progn (send *robot* :hand :arms :reset-pose) (send *ri* :hand-angle-vector (apply #\'concatenate float-vector (send *robot* :hand :arms :angle-vector))))')
        print('send hand_reset_pose command')
    def hand_hook_pose_cb(self, checked):
        eus_commmand_srv(command='(progn (send *robot* :hand :arms :hook-pose) (send *ri* :hand-angle-vector (apply #\'concatenate float-vector (send *robot* :hand :arms :angle-vector))))')
        print('send hand_hook_pose command')
    def hand_grasp_pose_cb(self, checked):
        eus_commmand_srv(command='(progn (send *robot* :hand :arms :grasp-pose) (send *ri* :hand-angle-vector (apply #\'concatenate float-vector (send *robot* :hand :arms :angle-vector))))')
        print('send hand_grasp_pose command')

    # Hrpsys Button Callback
    def hrpsys_start_abc_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :start-auto-balancer)')
        print('send hrpsys_start_abc command')
    def hrpsys_start_st_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :start-st)')
        print('send hrpsys_start_st command')
    def hrpsys_start_imp_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :start-impedance :arms)')
        print('send hrpsys_start_imp command')
    def hrpsys_stop_abc_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :stop-auto-balancer)')
        print('send hrpsys_stop_abc command')
    def hrpsys_stop_st_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :stop-st)')
        print('send hrpsys_stop_st command')
    def hrpsys_stop_imp_cb(self, checked):
        eus_commmand_srv(command='(send *ri* :stop-impedance :arms)')
        print('send hrpsys_stop_imp command')

    def shutdown_plugin(self):
        eus_commmand_srv.close()

    def save_settings(self, plugin_settings, instance_settings):
        # TODO save intrinsic configuration, usually using:
        # instance_settings.set_value(k, v)
        pass

    def restore_settings(self, plugin_settings, instance_settings):
        # TODO restore intrinsic configuration, usually using:
        # v = instance_settings.value(k)
        pass

    #def trigger_configuration(self):
        # Comment in to signal that the plugin has a way to configure
        # This will enable a setting button (gear icon) in each dock widget title bar
        # Usually used to open a modal configuration dialog
